File: WebSite/EulerAList/pset35.cpp
#include <iostream>     // std::cout
#include <algorithm>    // std::next_permutation, std::sort
#include <cmath>
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

unordered_set<long> primeList{2, 3, 5, 7};
vector<long> pm{2, 3, 5, 7};
string evenList = "02468";
vector<long> allList{2};

bool isPrime(long s)
{
    for (long i : pm)
    {
        if (i > sqrt(s)) return true;
        if (s % i == 0) return false;
    }
    return true;
}

long rotate(long s)
{
    long powTen = pow(10, (int)log10(s));
    int rem = s % 10;
    s = rem * powTen + s / 10;
    return s;
}

int main () 
{
    for (long i = 8; i < 1e6; i ++)
    {
        if (isPrime(i))
        {
            primeList.insert(i);
            pm.push_back(i);
        }
    }
    for (long i : pm)
    {
        string num = to_string(i);
        sort(num.begin(), num.end());
        bool isPossible = true;
        for (char c : evenList)
        {
            if (num.find(c) != string::npos)
            {
                isPossible = false;
                break;
            }
        }
        if (isPossible == false) continue;
        bool isCircular = true;
        long value = i;
        do 
        {
            value = rotate(value);
            if (primeList.find(value) == primeList.end())
            {
                isCircular = false;
                break;
            }

        } while (value != i);   
        if (isCircular || i == 2)
        {
            allList.push_back(i);
        }
    }
    for (int i : allList)
    {
        cout << i << ";";
    }
    cout << endl << allList.size() << endl;
}

File: WebSite/EulerAList/pset35.py
from math import sqrt, log10

pm = [2, 3, 5, 7]
primeList = set(pm)
evenList = "02468"
allList = [2]

def isPrime(s):
    for i in pm:
        if i > sqrt(s): 
            return True
        if s % i == 0:
            return False
    return True

def rotate(s):
    powTen = int(pow(10, int(log10(s))))
    rem = s % 10
    s = int(rem * powTen + s // 10)
    return s

for i in range(8, int(1e6)):
    if isPrime(i):
        primeList.add(i)
        pm.append(i)

for i in pm:
    num = str(i)
    num = sorted(num)
    isPossible = True
    for c in evenList:
        if c in num:
            isPossible = False
            break
    if not isPossible:
        continue
    isCircular = True
    value = i
    while True: 
        if not value in primeList:
            isCircular = False
            break
        value = rotate(value)
        if value == i:
            break   
    if isCircular or i == 2:
        allList.append(i)

for i in allList:
    print(i)
print(len(allList))